src/plots/cartesian/layout_attributes.js

```diff
@@ -465,6 +465,36 @@ module.exports = {
             'Only has an effect if `anchor` is set to *free*.'
         ].join(' ')
     },
+    categorymode: {
+        valType: 'enumerated',
+        values: [
+            'trace', 'category ascending', 'category descending', 'array'
```

Comment:
and maybe `categoryarray` instead of `categorylist`

@chriddyp we're already using the value `'array'` for [`tickmode`](https://github.com/plotly/plotly.js/blob/52be69a6b04d69a6c9ccfe9c6e4bb0e33cd03100/src/plots/cartesian/layout_attributes.js#L114).